codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java

```diff
@@ -156,14 +154,22 @@ public void channelActive(ChannelHandlerContext ctx) throws Exception {
 
     @Override
     public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        clientPrefaceString = clientPrefaceString(encoder.connection());
         // This handler was just added to the context. In case it was handled after
         // the connection became active, send the connection preface now.
         sendPreface(ctx);
     }
 
+    /**
+     * Releases the {@code clientPrefaceString}. Any active streams will be left in the
+     * open.
+     */
     @Override
     protected void handlerRemoved0(ChannelHandlerContext ctx) throws Exception {
-        dispose();
+        if (clientPrefaceString != null) {
```

Comment:
Are we guaranteed that this is called in all cases that lead to the channel shutting down?